### ISSUE TITLE
Reduce the packaged size of the library

### ISSUE THREAD

**Issue by Arachnid** [ensdomains/address-encoder#26]
As a result of pulling in many complete blockchain SDKs, this library has become rather bloated. In most cases, much of the code pulled in is extraneous, and only address encoding/decoding functionality is required. In others, one address encoder can be reused across multiple currencies that share an address encoding.

This issue is for a bounty for reducing the size of the library while preserving its functionality. For every halving in the size, 1 ether will be paid out, up to a maximum of 8 ether.

The exact formula for a bounty payout is:
```
payout = (log2(size_before_pr) - log2(size_after_pr)) * 1 ether
```

The current packaged size of the library can be determined by running `npm run size`, and this figure is what will be used to determine bounty payouts. As of writing, the initial figure is 6,411,165 bytes.

 - #27: 6,411,165 -> 2,983,989: 1.1034 ETH to @jay-dee7.
 - #28: 2,983,989 -> 1,956,940: 0.6086 ETH to @jay-dee7.
 - #30: 1,956,940 -> 738,548: 1.4058 ETH to @inspireme6. 
 - #31: 738,548 -> 440,007: 0.7472 ETH to @jay-dee7.
 - #37: 440,007 -> 275,979: 0.6730 ETH to @inspireme6.
 - #38: 275,979 -> 170,884: 0.6915 ETH to @maksim-tolo

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 8.0 ETH (2107.34 USD @ $263.42/ETH)  attached to it as part of the ensdomains fund.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4050).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ensdomains/address-encoder/26/4050).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $91,226.47 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 4 weeks, 1 day from now.
Please review their action plans below:


**1) [pbteja1998](https://gitcoin.co/pbteja1998) has started work.**

Clearly understood the issue. Will start working on it

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4050).

**Comment by Arachnid** [comment]
Now at 3,009,838 after #27. @jay-dee7 if you submit work, I will pay out 1.0909 Ether for your contribution.

**Comment by jay-dee7** [comment]
How do i resubmit work ? I'll raise another pr in few hours

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work for 8.0 ETH (2115.54 USD @ $264.44/ETH) has been submitted by__:

 1. [@jay-dee7](https://gitcoin.co/jay-dee7)



@Arachnid please take a look at the submitted work:
* [PR](https://github.com/ensdomains/address-encoder/pull/27) by @jay-dee7

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4050)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ensdomains/address-encoder/26/4050).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $91,293.31 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by Arachnid** [comment]
On my machine, current head is 2,983,989. I've updated the planned payout accordingly.

**Comment by jay-dee7** [comment]
I've updated the PR also added a check for size in travis config as well

**Comment by jay-dee7** [comment]
@Arachnid just a dummy question, what would i have to do, to be part of the ens organization?

**Comment by Arachnid** [comment]
@jay-dee7 Everyone who is part of the organisation has a full- or part- time paid contracting arrangement with us. We're definitely open to hiring more talented engineers, but our budget is pretty limited at the moment.

**Comment by x5engine** [comment]
hey, @Arachnid I have made a very good PR and hope it helps!

I got rid of what was taking almost 33% for nothing!

I hope I get paid asap thank you : )

ask me if there is any changes to do too!

there are more improvements to do too ( to make it smaller)

**Comment by jay-dee7** [comment]
@x5engine you've done a good job by splitting it, but can you please rebase your PR from master and update it? it seems we have merge conflicts and further you can replace `bs58check` with your `hextronweb` package and save a whooping `86KB`. PS: i removed TronWeb package all together in favor of `bs58check` but it seems that your package is just ~9 KB, so let's go with it.

**Comment by jay-dee7** [comment]
@Arachnid just as a confirmation, are we supposed to create separate packages? like specific npm package for this project?

**Comment by x5engine** [comment]
Hey @jay-dee7 thank you 😊

Well this is just a start and I can improve so much to make it so smaller and faster!

I think an independent package is a good way so it can be used in other packages but if it's not optimum we can have all the encoders in one place a folder here.

I'm really eager to get it a lot done for getting ETHs

So @Arachnid what do you think? Tell me and I can get this under 300kB ;)

**Comment by Arachnid** [comment]
@jay-dee7 Not sure what you mean? There's already an NPM package for this project; we're just looking for size reductions of the existing package.

**Comment by Arachnid** [comment]
Never mind, I understand.

If you need to publish a new package for this:
1. Please name it something like `[cryptocurrency]-address-encoder` so others can use it too.
2. Lock it to a specific published version in `package.json`.

**Comment by x5engine** [comment]
Okay sounds good I will do that and create a package with the correct name for the right encoder. and lock it on the package.json

**Comment by inspireme6** [comment]
Hey guys, I reduced the lib size by 62.27% just by removing the polkadot library :)

From 1,956,940 -> 738,548

**Comment by Arachnid** [comment]
@inspireme6 Amazing work! I've updated the list of pending payouts at the start.

**Comment by jay-dee7** [comment]
@arachind I've added a new PR with slight changes and further size reduction

**Comment by inspireme6** [comment]
What I did:

- Applied the latest changes from @jay-dee7 
- Forked the package "sha.js", removed all hashes except sha256 since it's the only one being used.
- Forked "crypto-add-codec" from @jay-dee7  and referenced the modified shajs package.
- Changed 'cashaddrjs' package for a more compact script which doesn't requires a biginteger lib.

436,035 -> 394,170 (9,61% decreased file size)

**Comment by Arachnid** [comment]
Really nice work from both of you! I've left some comments on #31 about some prerequisites before we can pull in a library with code from many places.

**Comment by x5engine** [comment]
@Arachnid  I got 1085093

https://github.com/ensdomains/address-encoder/pull/33

**Comment by x5engine** [comment]
![image](https://user-images.githubusercontent.com/12363020/75721722-23698d00-5cd9-11ea-8407-e1a7f9f949b1.png)
Bundle in windows is only 12KB

**Comment by Arachnid** [comment]
Merged #31, bringing the size to 440,007.

**Comment by jay-dee7** [comment]
@Arachnid i think size of the lib has been reduced by a fair amount so far and i'm not sure if i'll be able to participate actively anymore. (will always keep an eye tho).
By what time can i expect the bounty to be paid out?

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 2.45920  ETH (561.65 USD @ $228.39/ETH) has been granted to @inspireme6 for this issue from @Arachnid. ⚡️ 

Nice work @inspireme6! Your tip has automatically been deposited in the ETH address we have on file.

 * $76132.79 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 1.40580  ETH (321.07 USD @ $228.39/ETH) has been granted to @jay-dee7 for this issue from @Arachnid. ⚡️ 

Nice work @jay-dee7! Your tip has automatically been deposited in the ETH address we have on file.

 * $76132.79 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by Arachnid** [comment]
Thank you for your hard work! I've paid out for the work so far, and will reopen the bounty for any new optimisations.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__This Bounty has been completed.__

Additional Tips for this Bounty:
* Arachnid tipped 1.4058 ETH worth 321.07 USD to jay-dee7.
* Arachnid tipped 2.4592 ETH worth 561.65 USD to inspireme6.

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4050)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $74,305.68 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by jay-dee7** [comment]
> Thank you for your hard work! I've paid out for the work so far, and will reopen the bounty for any new optimisations.

@Arachnid i think you paid my reward to @inspireme6

**Comment by inspireme6** [comment]
> > Thank you for your hard work! I've paid out for the work so far, and will reopen the bounty for any new optimisations.
> 
> @Arachnid i think you paid my reward to @inspireme6

yea, I was going to say that lol. I'll tip you the whole amount from gitcoin.

**Comment by jay-dee7** [comment]
sure @inspireme6 🥇

**Comment by inspireme6** [comment]
> sure @inspireme6

okay just sent it @jay-dee7 :)

**Comment by Arachnid** [comment]
Doh, sorry! Thanks for sorting it out between you.

**Comment by inspireme6** [comment]
I'm dumb, I should've sent the difference, not the whole amount, tip me back the corresponding amount @jay-dee7 when you read this.

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 1.40000  ETH (319.74 USD @ $228.39/ETH) has been granted to @inspireme6 for this issue from @jay-dee7. ⚡️ 

Nice work @inspireme6! Your tip has automatically been deposited in the ETH address we have on file.

 * $74282.79 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by jay-dee7** [comment]
@inspireme6 i saw that lol, sent you the tip

**Comment by inspireme6** [comment]
Thanks @jay-dee7 @Arachnid

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__This issue now has a funding of 4.104 ETH (937.31 USD @ $228.39/ETH)  attached to it as part of the ensdomains fund.__

 * If you would like to work on this issue you can 'start work' [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4089).
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ensdomains/address-encoder/26/4089).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $75,220.10 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by x5engine** [comment]
Thank you 😊

Was a real pleasure and congrats to you guys @jay-dee7 @inspireme6 for you work

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work has been started__.


These users each claimed they can complete the work by 1 month from now.
Please review their action plans below:


**1) [davidbanu](https://gitcoin.co/davidbanu) has started work.**

Would love to continue to work on this...

Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4089).

**Comment by xiaods** [comment]
$ npm run size

> @ensdomains/address-encoder@0.1.3 size /Users/xiaods/Documents/Code/address-encoder
> browserify src/index.ts -p [tsify --noImplicitAny] | wc -c

440005

**Comment by gitcoinbot** [comment]
Issue Status: 1. **Open** 2. Started 3. Submitted 4. Done 

<hr>

__Work for 4.104 ETH (1000.82 USD @ $243.86/ETH) has been submitted by__:

 1. [@momodaka](https://gitcoin.co/momodaka)



@Arachnid please take a look at the submitted work:
* [PR](https://github.com/ensdomains/address-encoder/pull/35) by @momodaka

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4089)
* Want to chip in? Add your own contribution [here](https://gitcoin.co/issue/ensdomains/address-encoder/26/4089).
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $79,481.30 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)

**Comment by jay-dee7** [comment]
@momodaka @xiaods you guys realise that you won't be paid 4 eths for your submission? Please check the formula used for the payout calculation in the description. You can use `npm run size` or `yarn size` to compare size reduction and check your reward values

**Comment by xiaods** [comment]
@jay-dee7  thanks for your clarify, I don't need the reward, just sharp the codebase asap.

**Comment by inspireme6** [comment]
#37 reduces the size of the lib from 440007 to 275979 approximately. But it requires that the PR#2 of the cryptocodec gets merged first.

**Comment by Arachnid** [comment]
@inspireme6 Merged your PR #37 and updated the table!

**Comment by inspireme6** [comment]
> @inspireme6 Merged your PR #37 and updated the table!

Thanks @Arachnid! :tada:

**Comment by momodaka** [comment]
@jay-dee7 thanks for your mention,  I know the result. 
eliminates someline duplicate code

**Comment by maksim-tolo** [comment]
Hi @Arachnid! Could you please look at https://github.com/ensdomains/address-encoder/pull/38 and https://github.com/ensdomains/crypto-addr-serialize/pull/3?

**Comment by x5engine** [comment]
meanwhile, I got nothing from this.... so thanks guys

**Comment by Arachnid** [comment]
@maksim-tolo Sorry for the delay; I will take a look shortly.

@x5engine I'm sorry, but you never submitted a PR that passed the build.

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 0.67300  ETH (91.66 USD @ $136.19/ETH) has been granted to @inspireme6 for this issue from @Arachnid. ⚡️ 

Nice work @inspireme6! Your tip has automatically been deposited in the ETH address we have on file.

 * $80590.94 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by gitcoinbot** [comment]
⚡️ A tip worth 0.69150  ETH (94.18 USD @ $136.19/ETH) has been granted to @maksim-tolo for this issue from @Arachnid. ⚡️ 

Nice work @maksim-tolo! Your tip has automatically been deposited in the ETH address we have on file.

 * $80590.94 in Funded OSS Work Available at: https://gitcoin.co/explorer
 * Incentivize contributions to your repo: <a href='https://gitcoin.co/tip'>Send a Tip</a> or <a href='https://gitcoin.co/funding/new'>Fund a PR</a>
 * No Email? Get help on the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>

**Comment by Arachnid** [comment]
Thank you all for your hard work! I think we've got as far as we reasonably can, so I'm paying out the remaining bounties and closing the issue.

**Comment by gitcoinbot** [comment]
Issue Status: 1. Open 2. Started 3. Submitted 4. **Done** 

<hr>

__This Bounty has been completed.__

Additional Tips for this Bounty:
* Arachnid tipped 0.6915 ETH worth 94.18 USD to maksim-tolo.
* Arachnid tipped 0.6730 ETH worth 91.66 USD to inspireme6.
* jay-dee7 tipped 1.4000 ETH worth 190.67 USD to inspireme6.
* Arachnid tipped 1.4058 ETH worth 191.46 USD to jay-dee7.
* Arachnid tipped 2.4592 ETH worth 334.92 USD to inspireme6.

<hr>

* Learn more [on the Gitcoin Issue Details page](https://gitcoin.co/issue/ensdomains/address-encoder/26/4089)
* Questions? Checkout <a href='https://gitcoin.co/help'>Gitcoin Help</a> or the <a href='https://gitcoin.co/slack'>Gitcoin Slack</a>
* $80,145.71 more funded OSS Work available on the [Gitcoin Issue Explorer](https://gitcoin.co/explorer)
